### PR TITLE
Additional test cases added for backup/restore/delete_backup APIs

### PR DESCRIPTION
**Pull request by BA-Ramya**

### CODE REVIEW

File: examples/api/api_fam_backup.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/*
+ * api_fam_backup.cpp
+ * Copyright (c) 2020 Hewlett Packard Enterprise Development, LP. All rights
```
Comment (mashoodk) [inline examples/api/api_fam_backup.cpp:3]: 2022

File: examples/api/api_fam_backup.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        myFam->fam_abort(-1); // abort the program
+        // note that fam_abort currently returns after signaling
+        // so we must terminate with the same value
+        return -1;
+    }
+
+    // ... Initialization code here
+
+    try {
+        // create a 100 MB region with 0777 permissions and RAID5 redundancy
```
Comment (mashoodk) [inline examples/api/api_fam_backup.cpp:61]: Update comment - Region attributes are passed as NULL

File: src/memory_service/fam_memory_service_direct.cpp [hunk 3]
```diff
@@ -133,12 +133,19 @@ Fam_Memory_Service_Direct::Fam_Memory_Service_Direct(uint64_t memserver_id,
...
     struct stat info;
     if (stat(fam_backup_path.c_str(), &info) == -1) {
-        if ((errno == ENOENT) || (errno == ENOTDIR))
-            mkdir(fam_backup_path.c_str(),
-                  S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+        if ((errno == ENOENT) || (errno == ENOTDIR)) {
+            int ret = mkdir(fam_backup_path.c_str(),
+                            S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+            if (ret == -1)
+                cout << "Fam_Backup_Path Directory Creation failed. " << endl;
```
Comment (mashoodk) [inline src/memory_service/fam_memory_service_direct.cpp:147]: @faizan-barmawer Should we return error here ? May be we can ignore this error during memory server startup and throw error during fam_backup.

File: src/memory_service/fam_memory_service_direct.cpp [hunk 4]
```diff
@@ -133,12 +133,19 @@ Fam_Memory_Service_Direct::Fam_Memory_Service_Direct(uint64_t memserver_id,
 
     allocator =
         new Memserver_Allocator(num_delayed_free_Threads, fam_path.c_str());
-    fam_backup_path = (char *)strdup(config_options["fam_backup_path"].c_str());
+    std::string backup_location = config_options["fam_backup_path"];
+    std::size_t found = backup_location.find_last_of("/");
```
Comment (mashoodk) [inline src/memory_service/fam_memory_service_direct.cpp:137]: I did not understand this part of the code.

**Review by mashoodk — CHANGES_REQUESTED** [review]
I am yet to look at the permission scripts. Comments for example code and src changes

File: test/permission-test/fam-mm-permission-multi-user/t_backup_perm_440.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+
+
+# As user1:user1 Create a region with permission 777 of size 100MB
+su -c "./fam_create_region myRegion  100 777" user1 || test_failed "Create Region failed: $?" 1
+
+# As user1 : Allocate DataItem
+su -c "./fam_allocate_data myRegion myData 50 440" user1 || test_failed "Allocate DataItem user1 failed: $?"
+
+# As user2 : Backup DataItem
```
Comment (mashoodk) [inline test/permission-test/fam-mm-permission-multi-user/t_backup_perm_440.sh:30]: wrong comment. as user1

File: test/permission-test/fam-mm-permission-multi-user/t_backup_perm_440.sh [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ su -c "./fam_backup_data myRegion myData myBackup" user1 2>&1 1>/dev/null || test_failed "Backup DataItem user1 expected to pass: $?"
+
+# As user2 : Restore to DataItem
+! su -c "./fam_restore_backup_data myRegion myData myBackup" user1 || test_failed "Restore DataItem user1 expected to fail : $?"
+
+# As user2 : Delete backup
+! su -c "./fam_delete_backup_data myBackup" user1 || test_failed "Delete myBackup user1 expected to fail: $?"
+
+# As user2 : Backup DataItem
+! su -c "./fam_backup_data myRegion myData myBackup" user2 2>&1 1>/dev/null || test_failed "Backup DataItem user2 expected to pass: $?"
```
Comment (mashoodk) [inline test/permission-test/fam-mm-permission-multi-user/t_backup_perm_440.sh:40]: Is it expected to pass or fail ?

**Review by mashoodk — CHANGES_REQUESTED** [review]
Some more comments. Also can you add a comment on each test about expected behavior on a backup operation.

File: examples/api/api_fam_backup.cpp (hunk 1 above)
Comment (BA-Ramya) [reply examples/api/api_fam_backup.cpp]: Addressed

File: src/memory_service/fam_memory_service_direct.cpp (hunk 4 above)
Comment (BA-Ramya) [reply src/memory_service/fam_memory_service_direct.cpp]: This piece of code was initially added as stat of backup path fails if given backup is a file and not a folder. This code change has been reverted and the existence of fam_backup_path is done in fam_backup/fam_restore/fam_delete_backup APIs

File: test/permission-test/fam-mm-permission-multi-user/t_backup_perm_400.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+# As user1 : Allocate DataItem
+su -c "./fam_allocate_data myRegion myData 50 400" user1 || test_failed "Allocate DataItem user1 failed: $?"
+
+# As user1 : Backup DataItem
+ su -c "./fam_backup_data myRegion myData myBackup" user1 2>&1 1>/dev/null || test_failed "Backup DataItem user1 expected to pass: $?"
+
+# As user1 : Restore to DataItem
+! su -c "./fam_restore_backup_data myRegion myData myBackup" user1 2>&1 1>/dev/null || test_failed "Restore DataItem user1 expected to fail: $?"
+# As user1 : Delete backup
+! su -c "./fam_delete_backup_data myBackup" user1 2>&1 1>/dev/null || test_failed "Delete myBackup user1 expected to pass: $?"
```
[the remark this answers is not in the document]
Comment (BA-Ramya) [reply test/permission-test/fam-mm-permission-multi-user/t_backup_perm_400.sh]: Addressed. Backup Restore are passing with given permission, Delete with user1 is  failing. An issue #220 has been created

**Review by mashoodk — APPROVED** [review]

**Review by faizan-barmawer — APPROVED** [review]
Good to merge this code.
Need to revisit backup/restore permission tests.

Regards,
Faizan

**Review by mashoodk — APPROVED** [review]

**Pull request merged**